Add support for Ansible inventory file

dav-pascual opened the pull request:
Fix #3672
Implements #3796

Enables the generation of Ansible inventory file based on guest information. Supports predifined and default layout.

Enables support for running playbooks with this inventory file in common providers (container and local excluded)

Pull Request Checklist

* [ ] implement the feature
* [ ] write the documentation
* [ ] extend the test coverage
* [ ] update the specification
* [ ] adjust plugin docstring
* [ ] modify the json schema
* [ ] mention the version
* [ ] include a release note

dav-pascual commented:
TO-DO:

- Implement data container for ansible related data to keep consistency with the rest of the codebase?
- Add test coverage
- Add more details to #3672

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -3482,3 +3511,183 @@ def _run_action_phases(phases: list[Action]) -> tuple[list[ActionTask], list[Act
+            return groups
+
+        # Otherwise use role as group
+        if getattr(guest, 'role', None):
```
Every `Guest` instance has `role` attribute, no need for `getattr`.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -3482,3 +3511,183 @@ def _run_action_phases(phases: list[Action]) -> tuple[list[ActionTask], list[Act
+        """
+        return {'all': {'children': {'ungrouped': {}}}}
+
+    def _get_guest_ansible_config(self, guest: Guest) -> Optional[dict[str, Any]]:
```
Yeah, we will have this propagated from phase to `Guest` instances, just as `role` or any other attribute of a `provision` phase. This method will not be needed, simple `guest.ansible` will work. See `Guest.hardware` and `GuestData.hardware`, it's the same story.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -3482,3 +3511,183 @@ def _run_action_phases(phases: list[Action]) -> tuple[list[ActionTask], list[Act
+        :param guest: the guest to get variables for.
+        :returns: dictionary of Ansible host variables.
+        """
+        vars_dict: dict[str, Any] = {  # TODO check vars
```
Please, don't encode type in variable name, `vars` was good for the corresponding fmf key, it's good here as well.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -3482,3 +3511,183 @@ def _run_action_phases(phases: list[Action]) -> tuple[list[ActionTask], list[Act
+        vars_dict: dict[str, Any] = {  # TODO check vars
+            'ansible_host': guest.primary_address,
+            'ansible_user': getattr(guest, 'user', None),
+            'ansible_port': getattr(guest, 'port', None),
```
These two are available only when `guest` is an instance of `GuestSsh`, therefore you need `isinstance` first. Also, `GuestSsh` instance will have the attributes, always, check `GuestSsh` and `GuestSshData`. The attribute may be `None`, aka "use default value", but it will exist -> no need for `getattr()`.

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -3482,3 +3511,183 @@ def _run_action_phases(phases: list[Action]) -> tuple[list[ActionTask], list[Act
+        # Add any custom variables from provision step data's ansible.vars
+        ansible_config = self._get_guest_ansible_config(guest)
+        if ansible_config and 'vars' in ansible_config:
+            vars_dict.update(ansible_config['vars'])
```
This will become a simple `vars.update(guest.ansible.vars)` - `guest.ansible` will be a container, `guest.ansible.vars` will always be a dictionary, sometimes empty, true, but always a dictionary, no need for doubt and double-checking. We check once, on the boundary, when the raw user data enter tmt digestive tract :)

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -3482,3 +3511,183 @@ def _run_action_phases(phases: list[Action]) -> tuple[list[ActionTask], list[Act
+        else:
+            groups.append('ungrouped')
+
+        return groups
```
Right, once we get rid of dictionaries and the guessing they support, this will become much more elegant :)

```python
   def _get_host_groups(self, guest: Guest) -> list[str]:
        groups = ['all']  # All hosts are in 'all' group

        # Try to get ansible group from ansible.group key in provision guest data
        if guest.ansible.group:
            groups.append(guest.ansible.group)

        # Otherwise use role as group
        elif guest.role:
            groups.append(guest.role)

        else:
            groups.append('ungrouped')

        return groups
```

happz replied:
BTW `_get_host_vars` and `_get_host_groups` seem more like something `Guest` class should own, rather than the Ansible inventory generator. The set of variables or groups will not change no matter what the layout would be, they are pretty pure in this sense, depending only on what's in the guest itself. And it will allow easier composition of variables from different implementations of guests, without leaking outside. Something like this:

```python
@functools.cached_property
def ansible_host_vars(self) -> dict[str, Any]:
    return {
        'ansible_host': self.primary_address,
        **self.ansible.vars
    }

# And GuestSsh will override it:
@functools.cached_property
def ansible_host_vars(self) -> dict[str, Any]:
    return {
        **super().ansible_host_vars,
        'ansible_user': self.user,
        'ansible_port': self.port
    }

@functools.cached_property
def ansible_host_groups(self) -> list[str]:
    groups = ['all']  # All hosts are in 'all' group

    # Try to get ansible group from ansible.group key in provision guest data
    if self.ansible.group:
        groups.append(self.ansible.group)

    # Otherwise use role as group
    elif self.role:
        groups.append(self.role)

    else:
        groups.append('ungrouped')

    return groups
```

dav-pascual replied:
Fixed

LecrisUT commented on tmt/schemas/ansible.yaml:
```diff
@@ -0,0 +1,25 @@
+#
+
+$id: /schemas/ansible
+$schema: https://json-schema.org/draft-07/schema
```
Is this schema file linked to anything?

LecrisUT replied:
Do we need it at top-level? Doesn't this replace the multi-guest provision. I.e. what used to be multiple provision items can be a single ansible inventory?

happz replied:
This feature does not replace `provision` or multihost provisioning. It aims atvgap between tmt and more complex Ansible plays, by exposing what has been provisioned as Ansible inventory.

LecrisUT replied:
> Wouldn't that be a lot more work?

Last I've tried it, not really. It's orthogonal work for sure, but it is a single file implementation [^1]. Implementation wise, it would be the same, just moving the translation part into an ansible plugin layout. The plugin would still be run under the tmt environment so it would be fetching and translating the same data.

I would need to read the implementation in more details to see if it's more beneficial one way or another. Maybe a few things you could do with the plugin approach is to combine multiple sources, lazy evaluation, expand the inventory with subsequent steps, etc.

[^1]: at some point I looked into doing it for zerotier, I think was basing off of https://github.com/MyStarInYourSky/ansible-inventory-zerotier/blob/master/inventory_plugins/zerotier.py

happz replied:
Please, elaborate (not right now, of course...) how would tmt and the Ansible plugin work together to expose the existing tmt inventory. I'm curious, since I see it from the perspective of tmt invoking `ansible-playbook` - user wishing to set an extra fact for a particular guest, or enforce a Python interpreter for a subset of guests, that I can imagine from this PoV easily. But this PoV requires both shared and guest-specific keys.

dav-pascual replied:
@LecrisUT So if I understand correctly it would just be a matter of refactoring the code to use [python Inventory class](https://docs.ansible.com/ansible/latest/dev_guide/developing_inventory.html#developing-an-inventory-plugin) to build the inventory following the same logic.

Instead of a dict / yaml file we would have a python Inventory object?

For this PR I just followed the same philosophy that has worked for [mrack](github.com/neoave/mrack/blob/main/src/mrack/outputs/ansible_inventory.py) so I didn't stop much to evaluate that

dav-pascual replied:
I'll add here a comment that make a little bit clearer our usecase for this feat: https://github.com/teemtee/tmt/pull/3796#discussion_r2164679264 

`The main use case of this is having a lot of preexisting playbooks which target certain groups. All these playbooks should be run against the same inventory, which has a certain group hierarchy.`

LecrisUT replied:
Basically the core here is what is the role of creating the `inventory.yaml`. My understanding is that it is to make it easier for users to inject their own variables, configurations, etc. But if that's the case, why not allow the users to provide a `inventory.yaml` file directly. That is a more familiar language for them to interact with.

Ansible supports loading multiple inventory files, including the plugin file if I read the documentation correctly. The plugin approach would give us a lot of flexibility for the future, and it doesn't seem that difficult to implement.

Implementation wise, the plugin part would be very minimal, only providing the hosts defined in the `provision` key. Also any tmt specific facts like how to connect to the hosts.

If we allow the user to provide the `inventory.yaml` that would be read together with the plugin, we don't even need to define a logic for how to put each host in each group. The user would provide it themselves in the `inventory.yaml`, we just have to document the naming convention created for the hosts that are created by the plugin.

happz replied:
> Basically the core here is what is the role of creating the `inventory.yaml`. My understanding is that it is to make it easier for users to inject their own variables, configurations, etc. But if that's the case, why not allow the users to provide a `inventory.yaml` file directly. That is a more familiar language for them to interact with.

I'm sorry, but I don't follow here. How can user provide their own `inventory.yaml`, filled with host data, when at least their hostname/IP address is dynamic, known to tmt only after the guest has been provisioned? How is it connected?

> Ansible supports loading multiple inventory files, including the plugin file if I read the documentation correctly. The plugin approach would give us a lot of flexibility for the future, and it doesn't seem that difficult to implement.
> 
> Implementation wise, the plugin part would be very minimal, only providing the hosts defined in the `provision` key. Also any tmt specific facts like how to connect to the hosts.

Where would the plugin sit in the workflow of tmt and `ansible-playbook`? How would the _Ansible inventory plugin_ get the relevant information from tmt? This is the piece I am missing completely: Ansible does support custom inventory plugins, but these plugins query some kind of infrastructure and present it to Ansible; how will the inventory plugin get the hostname, name and role of a particular guest so it could be included in the inventory presented to Ansible?

LecrisUT replied:
Oh, so you take the user-provided `inventory.yaml` (if any) and then you rewrite it with the tmt facts? That sounds like a good approach. But then the `python_interpreter`, `group_vars`, etc. do not need to be exposed.

I am not a fan of having `ansible` as a top-level key. Preferably this would be under `provision`. I think it's possible with a "guestless" provision, something like
```yaml
provision:
  - name: server
    how: virtual
  - name: client
    how: container
  - name: dummy_not_really_a_guest
    how: ansible
    file: path/to/inventory.yaml
```

If a separate how ansible is not provided, then the step can just be injected with a fresh `inventory.yaml`. One issue might be the order of the other provision steps, but maybe you can delay the execution until the prepare step consumes it.

@happz @psss wdyt? Wouldn't this interface be cleaner

dav-pascual replied:
@LecrisUT `python_interpreter` and `group_vars` are only set if the user provides them. It's an optional feature, for the case the user wants to customize them. The inventory is always generated, even if no layout is provided, using a default basic layout. Take a look at: https://github.com/teemtee/tmt/pull/3796 

I discussed in the design where to put the ansible key with @happz and @psss . Somehow under provision was also my first thought too, but since multihost provision spec is a list, the choices are to modify multihost provision API or put ansible as a top-level key. The former would break many current fmf plan files, so the latter is the best choice. The inventory is used by multiple phases to execute the playbooks so it makes sense IMO

happz replied:
Let's make one thing clear: there is no "singlehost" vs "multihost", or "multihost API" - `provision` (or `prepare` or `report`) is *always* a list. Even when it's written down as keys directly under `provision`, internally it's a list, and it's always treated as a list of guests to provision. Or a list of report plugins to invoke. So the focus should be on the list form, the simplified, "singlehost" form is just a syntax sugar for users.

> I discussed in the design where to put the ansible key with @happz and @psss . Somehow under provision was also my first thought too, but since multihost provision spec is a list, the choices are to modify multihost provision API or put ansible as a top-level key. The former would break many current fmf plan files, so the latter is the best choice. The inventory is used by multiple phases to execute the playbooks so it makes sense IMO

Correct, it was mentioned during the discussion. The implementation in the core would be indeed similar, iterate over `provision`, pick up the special `how: ansible` and render the inventory. However, there are more places that iterate over "guests", so it's probably not a question of *an*  `isinstance()` call, but rather multiple such calls in several places. But leaving that aside, I for one prefer the dedicated `ansible` block over special phases because I believe it makes for better, more readable and consistent plan specification. And I never liked lists of multiple types when some types have special handling. Therefore the special `how: ansible` phase gets -1 from me, and I voted for `ansible` key on plan level when the idea was discussed.

dav-pascual replied:
@LecrisUT I agree with you on `python_interpreter` part. We can use group vars  in inventory.yaml to set this up. I'll remove this part. 

As for the user provided vars for individual I still think there are usefull and necessary because you cannot define dynamic host data in the inventory. This is part of tmt work of gathering host data from the plan spec. 
So a user may want to define custom vars in the plan to customize a host vars, not having to worry about the inventory, it should be abstracted from the inmediate user for some cases.

dav-pascual replied:
As discussed, I removed python_interpreter and kept host vars in the metadata

LecrisUT commented on tmt/steps/provision/__init__.py:
```diff
@@ -3235,6 +3249,19 @@ def load(self) -> None:
+            layout = self.plan.node.get('ansible', {}).get('inventory', {}).get('layout')
+
+            inventory = self._ansible_inventory.generate(self.guests, layout)
+            inventory_path = Path('inventory.yaml')
```
One issue that we were considering at some point is where to put the `inventory.yaml` and other `host_vars`, etc. when those can be exposed. I believe we wanted it under some `TMT_PLAN_DATA` path so that it is accessible for each `prepare` step.

dav-pascual replied:
This is something I wondered while coding the feature. For now, the path logic and order is 
1. Relative to fmf file
2. Current working dir

psss replied:
Let's keep the logic consistent with the [ansible playbook path](https://tmt.readthedocs.io/en/stable/plugins/prepare.html#configuration) handling, that is it should relative to the `fmf` metadata root.

dav-pascual replied:
@happz @psss thoughts?

dav-pascual replied:
@LecrisUT This would be also a read-only file, as we are referring to the layout inventory. The output inventory path is fixed and always goes to workdir, in provision folder.

dav-pascual replied:
I'm using write utils function which writes to the workdir to avoid duplicating code, that's why I don't define the full path at the beginning. I have renamed the variable to make it less confusing

LecrisUT reviewed (commented):
Thanks for working on this. Just to gauge the scope of this, will allowing a `host_vars`/`group_vars` organization instead of the `ansible.vars` be out-of-scope for this PR for now?

dav-pascual commented:
I updated with the following changes:
- Took Ansible related classes and methods to a separate file (and some methods to Guest class)
- Created Ansible Data container classes
- Added python_interpreter option in ansible root level ansible.inventory key. This sets ansible_python_interpreter key ([ref](https://docs.ansible.com/ansible/latest/inventory_guide/intro_inventory.html#id21))

dav-pascual commented:
> Thanks for working on this. Just to gauge the scope of this, will allowing a `host_vars`/`group_vars` organization instead of the `ansible.vars` be out-of-scope for this PR for now?

The idea is to start simple to enable the support of inventory file. From there anyone can iterate and enhance the feature.

Adding native support of `host_vars`/`group_vars` will be nice too, but I think it should be done in a different PR.

dav-pascual commented:
Updated:
- Fixed pre-commit issues
- Changed layout path logic to fmf metadata root
- Removed python_interpreter key
- Added mapping ansible -> guest_ansible due to shadowing error by existing "ansible()" method

Next:
- Add test coverage
- Review JSON schema
- Review docstring and documentation

psss commented:
Thanks much for preparing this! Proposing for the next sprint.

LecrisUT commented on tests/provision/ansible-inventory/test.sh:
```diff
@@ -0,0 +1,105 @@
+        rlRun -s "${tmt_command} ${planName}" 0
+
+        # Check that inventory file was generated in the provision step and directory
+        inventory_file="$(find $run -path '*/provision/inventory.yaml' | head -1)"
```
Yes, if the file structure changes we should be aware of it and see if it's sensible. E.g. it could hide the path changing from plan_workdir to run_workdir

dav-pascual replied:
Updated

dav-pascual commented on tmt/ansible.py:
```diff
@@ -0,0 +1,322 @@
+        if layout_path:
+            resolved_path = self._plan.anchor_path / layout_path
+            try:
+                return tmt.utils.yaml_to_dict(resolved_path.read_text())
```
'ungrouped' is added when is identified as a group that doesn't exist.

But your comment made me realize there was a need for normalization method, the layout could come with implicit default groups 'all' and 'ungrouped'. It would fail for the former.

I have update to include such method and ensure both groups are present explicitely

happz commented on tmt/steps/provision/__init__.py:
```diff
@@ -2609,21 +2662,34 @@ def _run_ansible(
         parent = cast(Provision, self.parent)
 
+        provision_step = parent.plan.provision
+        inventory_path = provision_step.workdir / 'inventory.yaml'  # type: ignore[operator]
```
1. `provision_step` does not need to exist, `inventory_path = parent.plan.provision.workdir / 'inventory.yaml'` would as nicely.
2. `ignore[operator]` is suspicious, why is it needed? It suggests there is something unexpected happening when creating the path.

happz commented on docs/releases.rst:
```diff
@@ -15,6 +15,12 @@ While running ``tmt plan export``, the exported plan now includes
 combined context from multiple sources, including command line,
 making it consistent with ``tmt plan show`` command.
 
+TMT now supports automatic generation of Ansible inventory files
```
```suggestion
tmt now supports automatic generation of Ansible inventory files
```

happz commented on docs/releases.rst:
```diff
@@ -85,7 +91,6 @@ the environment provided by mock and shell access for executing tests.
 This feature is currently experimental and subject to change. We are
 looking forward to user feedback.
 
-
```
Unrelated change, and a removal of a very fine empty line.

happz commented on tmt/ansible.py:
```diff
@@ -0,0 +1,336 @@
+        :param inventory: the inventory dictionary to modify.
+        :param guest: the guest to add to the inventory.
+        """
+        host_vars = guest.ansible_host_vars
```
`host_vars` exists only to be assigned to `inventory['all']['hosts'][guest.name]`, not needed.

dav-pascual commented on tmt/steps/provision/__init__.py:
```diff
@@ -3428,6 +3507,25 @@ def load(self) -> None:
+            self.write(inventory_path, tmt.utils.dict_to_yaml(inventory))
+            self.info('ansible', f'Inventory saved to {inventory_path}')
+        except tmt.utils.FileError as exc:
+            self.debug(f"Failed to save Ansible inventory: {exc}")
```
This is meant to log the message, not raise an exception. We don't want overall runs that don't use the inventory to fail because of it IMO

happz replied:
Ah, I misread that completely.

> We don't want overall runs that don't use the inventory to fail because of it IMO

Hmm, why not? Failed to save a file, isn't it expected to exist e.g. when `prepare/ansible` runs later?

dav-pascual replied:
That sounds like a good approach to me. Only thing is that we would change the logic of always generating the inventory file in provision (in save method) to generating it on demand (e.g. when accessed from prepare phase with playbooks). I cannot think of any issue with that right now

dav-pascual replied:
I have added a new commit with this approach. Check it out

dav-pascual commented on tmt/steps/provision/__init__.py:
```diff
@@ -2609,21 +2662,34 @@ def _run_ansible(
+            playbook,
+        ]
+
+        ansible_command += Command(*cmd_args)
```
updated

happz commented:
/packit build

LecrisUT commented on tmt/ansible.py:
```diff
@@ -0,0 +1,322 @@
+    )
+
+    @classmethod
+    def from_spec(cls, spec: Any) -> 'GuestAnsible':
```
You can check `FmfId.from_spec` for a reference. Basically to/from spec should expect a dict type and the `TypedDict` there is just a fancy form of that dict which can restrict the keys and value types it should have. The main thing that you should account for these if the presence of default values and how to handle them, e.g. if you should save a default value or not include the value when saving it etc. Here are two options where you either save the nul values in the spec or not
```python
class _RawGuestAnsible(TypedDict)
    group: Optional[str]
    vars: dict[str, Any]

@container
class GuestAnsible(SerializableContainer):
    @classmethod
    def from_spec(cls, spec: _RawGuestAnsible) -> "GuestAnsible":
        return cls(**spec)

    def to_spec(self) -> _RawGuestAnsible:
        return self.to_dict()
```
```python
class _RawGuestAnsible(TypedDict, total=False)
    group: str
    vars: dict[str, Any]

@container
class GuestAnsible(SerializableContainer):
    @classmethod
    def from_spec(cls, spec: _RawGuestAnsible) -> "GuestAnsible":
        return cls(**spec)

    def to_spec(self) -> _RawGuestAnsible:
        spec_dict = self.to_dict()
        # Sanitize some values
        if not self.group:
            del spec_dict["group"]
        if not self.vars:
            del spec_dict["vars"]
        return spec_dict
```

dav-pascual replied:
added commit with suggestion, check it out

LecrisUT replied:
Yep, looks fine, just a minor comment added.

LecrisUT commented on tmt/steps/provision/__init__.py:
```diff
@@ -1326,6 +1342,8 @@ def get_data_class(cls) -> type[GuestData]:
 
     environment: tmt.utils.Environment
 
+    ansible: GuestAnsible
```
```suggestion
    ansible: Optional[GuestAnsible]
```
?

dav-pascual commented on tmt/steps/provision/local.py:
```diff
@@ -79,6 +79,7 @@ def _run_ansible(
         playbook = self._sanitize_ansible_playbook_path(playbook, playbook_root)
 
         try:
+            # TODO: add ansible inventory support
```
Comment added and issue created: https://github.com/teemtee/tmt/issues/4216

LecrisUT commented on tmt/ansible.py:
```diff
@@ -82,34 +79,30 @@ class GuestAnsible(SerializableContainer):
-        raise tmt.utils.SpecificationError(f"Invalid Ansible specification: {spec}")
+        return cls(
+            group=spec.get('group'),
+            vars=spec.get('vars') or {},
```
```suggestion
            vars=spec.get('vars', {}),
```
You could also `cls(**spec)`. If it has extra keys that it would make it fail, that's something we should know.

dav-pascual replied:
mypy/pyright was complaining that None case was not handled for both suggestions. I added an ignore comment

LecrisUT commented on tmt/ansible.py:
```diff
@@ -0,0 +1,308 @@
+    )
+
+    @classmethod
+    def from_spec(cls, spec: Union[_RawGuestAnsible, None]) -> 'GuestAnsible':
```
```suggestion
    def from_spec(cls, spec: Optional[_RawGuestAnsible]) -> 'GuestAnsible':
```
Equivalent, but we use this convention. Similar below

tcornell-bus commented on docs/releases.rst:
```diff
@@ -19,6 +19,12 @@ New :ref:`sprints` section has been added to the documentation for
+during the provision phase. The new :ref:`/spec/plans/ansible`
+configuration allows users to define custom inventory layouts and
+organize provisioned guests into Ansible groups with host-specific
+variables. See the :ref:`/spec/plans/ansible` and
```
This comes out to say "See the ansible and ansible sections for configuration details."
Maybe adding "provision" before the second ansible will make it clear they are different.


```suggestion
variables. See the :ref:`/spec/plans/ansible` and provision
```
